proxy/src/connection.rs

```diff
@@ -252,24 +262,47 @@ impl Future for Connecting {
 
     fn poll(&mut self) -> Poll<Self::Item, Self::Error> {
         loop {
-            *self = match self {
-                Connecting::Plaintext { connect, tls } => {
+            self.state = match &mut self.state {
+                ConnectingState::Plaintext { connect, tls } => {
+                    trace!("Connecting: state=plaintext; tls={:?};",tls);
                     let plaintext_stream = try_ready!(connect.poll());
                     set_nodelay_or_warn(&plaintext_stream);
                     match tls.take().expect("Polled after ready") {
                         Conditional::Some(config) => {
-                            let upgrade_to_tls = tls::Connection::connect(
+                            trace!("plaintext connection established; trying to upgrade");
+                            let upgrade = tls::Connection::connect(
                                 plaintext_stream, &config.identity, config.config);
-                            Connecting::UpgradeToTls(upgrade_to_tls)
+                            ConnectingState::UpgradeToTls(upgrade)
                         },
                         Conditional::None(why) => {
+                            trace!("plaintext connection established; no TLS ({:?})", why);
                             return Ok(Async::Ready(Connection::plain(plaintext_stream, why)));
                         },
                     }
                 },
-                Connecting::UpgradeToTls(upgrading) => {
-                    let tls_stream = try_ready!(upgrading.poll());
-                    return Ok(Async::Ready(Connection::tls(BoxedIo::new(tls_stream))));
+                ConnectingState::UpgradeToTls(upgrade) => {
+                    match upgrade.poll() {
+                        Ok(Async::NotReady) => return Ok(Async::NotReady),
+                        Ok(Async::Ready(tls_stream)) => {
+                            let conn = Connection::tls(BoxedIo::new(tls_stream));
+                            return Ok(Async::Ready(conn));
+                        },
+                        Err(e) => {
+                            warn!(
+                                "TLS handshake with {:?} failed: {}\
+                                    -> falling back to plaintext",
+                                self.addr, e,
+                            );
+                            let connect = TcpStream::connect(&self.addr);
+                            // TODO: emit a `HandshakeFailed` telemetry event.
```

Comment:
Not a blocker: I don't understand if this means that the "is tls" telemetry will be reported correctly for connections that fall back or not. It seems like the current telemetry reports if we are attempted TLS, not if the connection actually is TLS, but maybe I'm just misunderstanding it. I would love to chat about this after this is merged.

Reply:
I've done some testing and it seems that you're right, the label is not added correctly. I'm working on addressing that, but I'm going to continue to do so in another branch so that this can be merged.